src/radical/pilot/raptor/master.py

```diff
@@ -427,105 +440,201 @@ def join(self):
+
+
+    # --------------------------------------------------------------------------
+    #
+    def _request_cb(self, tasks):
+
+        tasks = ru.as_list(tasks)
+
+        self._log.debug('request_cb: %d', len(tasks))
 
         try:
-            filtered = self.request_cb(requests)
+            filtered = self.request_cb(tasks)
             if filtered:
-                self.request(filtered)
+                for task in filtered:
+                    self._log.debug('REQ cb: %s' % task['uid'])
+                self.submit_tasks(filtered)
 
         except:
             self._log.exception('request cb failed')
-            # FIXME: fail the request
+            # FIXME: fail tasks
 
 
+    # --------------------------------------------------------------------------
+    #
+    def request_cb(self, tasks):
+
+        # FIXME: document task format
+        return tasks
 
 
     # --------------------------------------------------------------------------
     #
-    def _result_cb(self, msgs):
+    def _result_cb(self, tasks):
 
-        for msg in ru.as_list(msgs):
+        tasks = ru.as_list(tasks)
+
+        try:
+            self.result_cb(tasks)
+
+        except:
+            self._log.exception('result callback failed')
+
+        for task in tasks:
+
+            # check if the task was submited via the task_service EP
+            tid = task['uid']
+
+            if tid in self._task_service_data:
+                # update task info and signal task service thread
+                self._log.debug('unlock 1 %s', tid)
+                self._task_service_data[tid][1] = task
+                self._task_service_data[tid][0].set()
 
-            # update result and error information for the corresponding request UID
-            uid = msg['req']
-            out = msg['out']
-            err = msg['err']
-            ret = msg['ret']
-            val = msg['val']
+            ret = int(task.get('exit_code', -1))
 
-            req = self._requests[uid]
-            req.set_result(out, err, ret, val)
+            if ret == 0: task['target_state'] = rps.DONE
+            else       : task['target_state'] = rps.FAILED
 
-            try:
-                self.result_cb(req)
-            except:
-                self._log.exception('result callback failed')
+        self.advance(tasks, rps.AGENT_STAGING_OUTPUT_PENDING,
+                            publish=True, push=True)
 
-            # if the request is a task, also push it into the output queue
-            if req.task:
 
-                req.task['stdout']    = out
-                req.task['stderr']    = err
-                req.task['exit_code'] = ret
+    # --------------------------------------------------------------------------
+    #
+    def result_cb(self, tasks):
 
-                if ret == 0: req.task['target_state'] = rps.DONE
-                else       : req.task['target_state'] = rps.FAILED
+        # FIXME: document task format
```

Comment:
* Are tasks dictionaries or Task instances?
* What fields will be added or changed in the Task?
* Where/how does the 4-element tuple of Worker executor output appear?

Reply:
> Are tasks dictionaries or Task instances?

`Task` instances only really exist in the client API.  Throughout the internal code we use dictionaries to represent those task instances, but those dicts are not really `Task` instances themself.

> What fields will be added or changed in the Task?

Sorry for being slow on the documentation for this.  It is a bit unclean that we expose our internal task dictionary representation to the application via the callbacks and I am a bit hesitant to document all internal data fields (applications should not rely on those).  We will introduce a more formal `TaskDict` class which encapsulates the public part of that representation along with documentation - but that will only arrive in 1.15 I'm afraid.

> Where/how does the 4-element tuple of Worker executor output appear?

The `(stdout, stderr, return_code, return_value, exception)` 5-tuple (yes, it grew) will be available on the task dictionary and also on the API level `Task` instances, under the same names: `'stdout', 'stderr', 'exit_code', 'return_value',` and `'exception'`